src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/collator/Collator.java

```diff
@@ -0,0 +1,148 @@
+   */
+  private static final int collatorFrequency = 1; // Measured in terms of number of evaluationIntervalPeriods
+
+  private static final int evalIntervalSeconds = 5;
+
+  private final List<Decider> deciders;
+
+  private final ActionGrouper actionGrouper;
+
+  public Collator(Decider... deciders) {
+    this(new SingleNodeImpactActionGrouper(), deciders);
+  }
+
+  public Collator(ActionGrouper actionGrouper, Decider... deciders) {
+    super(evalIntervalSeconds, collatorFrequency);
+    this.deciders = Arrays.asList(deciders);
+    this.actionGrouper = actionGrouper;
+  }
+
+  @Override
+  public String name() {
+    return NAME;
+  }
+
+  /**
+   * The collator uses an action grouping strategy to first group actions by instanceIds. Then, the
+   * collator polarizes the list of actions per instance to be in the same direction of pressure,
+   * i.e. all the polarized actions either increase pressure on a node, or decrease pressure on a
+   * node.
+   *
+   * <p>When there are conflicting actions suggested by the deciders for an instance, the
+   * polarization logic prefers pruning actions that decrease stability retaining only those that
+   * increase stability. </p>
+   *
+   * @return A {@link Decision} instance that contains the list of polarized actions.
+   */
+  @Override
+  public Decision operate() {
+    final List<Action> proposedActions = getAllProposedActions();
+    final Map<NodeKey, List<Action>> actionsByNode = actionGrouper
+        .groupByInstanceId(proposedActions);
+    final List<Action> prunedActions = new ArrayList<>();
+    actionsByNode.forEach((nodeKey, actions) -> prunedActions.addAll(polarize(nodeKey, actions)));
+
+    final Decision finalDecision = new Decision(System.currentTimeMillis(), NAME);
+    finalDecision.addAllActions(prunedActions);
+    return finalDecision;
+  }
+
+  @NonNull
+  private List<Action> getAllProposedActions() {
+    final List<Action> proposedActions = new ArrayList<>();
+    if (deciders != null) {
+      for (final Decider decider : deciders) {
+        List<Decision> decisions = decider.getFlowUnits();
+        if (decisions != null) {
+          decisions.forEach(decision -> {
+            if (decision.getActions() != null) {
+              proposedActions.addAll(decision.getActions());
+            }
+          });
+        }
+      }
+    }
+    return proposedActions;
+  }
+
+  private List<Action> polarize(final NodeKey nodeKey, final List<Action> actions) {
+    final List<Action> pressureIncreasingActions = new ArrayList<>();
+    final List<Action> pressureNonIncreasingActions = new ArrayList<>();
+
+    for (final Action action : actions) {
+      ImpactVector impactVector = action.impact().getOrDefault(nodeKey, new ImpactVector());
+
+      // Classify the action as pressure increasing action if the impact for any dimension is
+      // increasing pressure.
+      if (impactVector.getImpact()
```

Comment:
This classifies an increase across any dimension as increasing pressure. Shouldn't we check on a per dimension basis? Otherwise there is no value in having different dimensions in the impact vector.

Suppose you have an action to offload some data from heap to a file on disk. This would reduce heap but increase disk pressure. If there is no other action trying to decrease disk pressure, then the overall disk on the node is not under contention and we should let this action go through.

Reply:
Makes sense, I assumed for the first iteration we would not have such actions and had not thought of polarizing at a dimension level. I will change it.